### PR TITLE
Add base spec for taste-tester for ssh

### PR DESCRIPTION
**Pull request by nishchintraina**
Set up a baseline for testing in tt.
Add spec test for testing building of ssh command

### CODE REVIEW

**Comment by nishchintraina** [comment]
@jaymzh @NaomiReeves - tunnel spec's coming in another PR :)

File: spec/ssh_spec.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+# limitations under the License.
+
+require 'logger'
+require 'base64'
+require 'between_meals/util'
+require 'taste_tester/logging'
+require 'taste_tester/exceptions'
+require 'taste_tester/ssh_util'
+require 'taste_tester/ssh'
+require 'taste_tester/config'
```
Comment (jaymzh) [inline spec/ssh_spec.rb:24]: you should change all these to:
```
require_relative '../lib/taste_tester/logging'
...
```

so the tests run against the repo and not installed code.

File: spec/ssh_spec.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+describe TasteTester::SSH do
+  let(:logger) do
+    Logger.new('/dev/null')
+  end
+
+  tt_ssh = TasteTester::SSH.new('mock_host')
```
Comment (jaymzh) [inline spec/ssh_spec.rb:31]:
```ruby
  let(:tt_ssh) do
    TasteTester::SSH.new('mock_host')
  end
```

File: spec/ssh_spec.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    it 'test build ssh command' do
+      expect(tt_ssh.build_ssh_cmd('mock_ssh', ['cmd1', 'cmd2'])).to eq(
+        'mock_ssh root@mock_host ' \
+        "\"echo 'Y21kMSAmJiBjbWQy' | base64 --decode | bash -x\"",
+      )
+    end
+  end
+
+  context 'test custom configs linux' do
+    before(:each) do
```
Comment (jaymzh) [inline spec/ssh_spec.rb:60]: why before _each_ ? Unless you're changing these in tests, you can just do `before do`

File: spec/ssh_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    it 'test ssh target' do
+      expect(tt_ssh.ssh_target).to eq(
+        'root@mock_host',
+      )
+    end
+
+    it 'test build ssh command' do
+      expect(tt_ssh.build_ssh_cmd('mock_ssh', ['cmd1', 'cmd2'])).to eq(
+        'mock_ssh root@mock_host ' \
```
Comment (jaymzh) [inline spec/ssh_spec.rb:53]: backslashes are python-y. Use `+`

File: spec/ssh_spec.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    it 'test ssh base command' do
+      expect(tt_ssh.ssh_base_cmd).to eq(
+        'mock_ssh_command ' \
+        '-J mock_jump_user@mock_jump_host ' \
+        '-T ' \
+        '-o BatchMode=yes ' \
+        '-o UserKnownHostsFile=/dev/null ' \
+        '-o StrictHostKeyChecking=no ' \
+        '-o ConnectTimeout=10',
```
Comment (jaymzh) [inline spec/ssh_spec.rb:75]: Use + instead of \

**Review by jaymzh — CHANGES_REQUESTED** [review]
This PR makes me *HAPPY*! Nice work.

Few minor things below, but that's it.

I was gonna say you have to actually modify the GH stuff so they run... but turns out they already do! Neat.

**Comment by nishchintraina** [comment]
> This PR makes me _HAPPY_! Nice work.
> 
> Few minor things below, but that's it.
> 
> I was gonna say you have to actually modify the GH stuff so they run... but turns out they already do! Neat.

Thanks @jaymzh! I will follow up with the suggested fixes in a bit. <3

n00n question - I was trying to also test the exception message for failures, but the `error!` block inside `SSH:Util` calls a fail - and it doesn't get caught by the following
```
    it 'test ssh exception message' do
      expect(tt_ssh.error!).to raise_error(
        TasteTester::Exceptions::SshError
      )
    end
```

Any ideas?

**Comment by jaymzh** [comment]
> n00b question - I was trying to also test the exception message for failures, but the `error!` block inside `SSH:Util` calls a fail - and it doesn't get caught by the following
> 
> ```
>     it 'test ssh exception message' do
>       expect(tt_ssh.error!).to raise_error(
>         TasteTester::Exceptions::SshError
>       )
>     end
> ```
> 
> Any ideas?

Yeah - you can't use `()`, you need a code-block to catch the error.

```ruby
expect { tt_ssh.error! }.to raise_error(...)
```

**Comment by nishchintraina** [comment]
>     it 'test ssh exception message' do
>       expect(tt_ssh.error!).to raise_error(
>         TasteTester::Exceptions::SshError
>       )
>     end

Thanks! I think it's looks good-y now. :)

**Review by jaymzh — APPROVED** [review]
I appreciate the fact that you added the crazy windows target stuff in there too.

**Comment by jaymzh** [comment]
You need to fix your lint/tests before I can merge though.

**Comment by nishchintraina** [comment]
Huh yeah I was looking into this - I think there is a problem with how I am overriding the tt config - it breaks the tunnel tests - both pass individually but not together. :(

File: spec/ssh_spec.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      )
+    end
+  end
+
+  context 'test custom configs linux' do
+    before do
+      TasteTester::Config.ssh_connect_timeout 10
+      TasteTester::Config.jumps 'mock_jump_user@mock_jump_host'
+      TasteTester::Config.ssh_command 'mock_ssh_command'
+      TasteTester::Config.user 'rossi'
```
Comment (nishchintraina) [inline spec/ssh_spec.rb:79]: @jaymzh  This is mutating shit globally, so it pollutes other tests. Sorry :( Once I find a way to do this I will fix the other test too.

Comment (jaymzh) [reply spec/ssh_spec.rb:79]: Oh that makes sense, it's a singleton.

You can save the values, and then use an `after` block to restore them.

Comment (nishchintraina) [reply spec/ssh_spec.rb:89]: Thanks! I was able to figure it out using the save/restore context in `Mixlib::Config` :)

**Comment by jaymzh** [comment]
Please don't merge... rebase.

**Comment by nishchintraina** [comment]
> Please don't merge... rebase.

I... don't git it. :(

I mean I do, but this time, I didn't. 

Anyway, should be good now. :)

**Review by jaymzh — APPROVED** [review]
Yay tests!

**Pull request merged**